src/javascript/note.js

```diff
@@ -1,94 +1,215 @@
+    if (this.percentage < 0.05) {
+      this.vm.result.marks.perfect += 1;
+      this.vm.markJudge = "Perfect";
+    } else if (this.percentage < 0.25) {
+      this.vm.result.marks.good += 1;
+      this.vm.markJudge = "Good";
+    } else if (this.percentage < 0.5) {
+      this.vm.result.marks.offbeat += 1;
+      this.vm.markJudge = "Offbeat";
+    }
+  }
 
-  this.hitAndCountScore = function () {
+  hitAndCountScore(isHolding) {
     this.vibrate(25);
-    const percentage = this.getDiffPercentage();
+    this.judge();
+    const percentage = this.percentage;
     let accuracyPercent = 100 * (1 - percentage);
-    vm.result.totalPercentage += accuracyPercent;
-    vm.result.totalHitNotes += 1;
-    vm.result.score += 3 * accuracyPercent;
-    vm.result.combo += 1;
-    vm.result.maxCombo =
-      vm.result.combo > vm.result.maxCombo
-        ? vm.result.combo
-        : vm.result.maxCombo;
-    if (percentage < 0.05) {
-      vm.result.marks.perfect += 1;
-      vm.markJudge = "Perfect";
-    } else if (percentage < 0.25) {
-      vm.result.marks.good += 1;
-      vm.markJudge = "Good";
-    } else if (percentage < 0.5) {
-      vm.result.marks.offbeat += 1;
-      vm.markJudge = "Offbeat";
+    if (!isHolding) {
+      // hold note does not count towards accuracy percent
+      this.vm.result.totalPercentage += accuracyPercent;
+      this.vm.result.totalHitNotes += 1;
     }
-    this.hitIndicator(vm);
-  };
-
-  this.isOutOfCanvas = function () {
-    const isOut = this.y > canvas.height;
-    if (vm.started && isOut) {
-      vm.result.marks.miss += 1;
-      vm.result.totalHitNotes += 1;
-      vm.result.combo = 0;
-      vm.markJudge = "Miss";
+    this.vm.result.score += 2 * accuracyPercent * this.vm.fever.value;
+    this.vm.result.combo += 1;
+    this.vm.result.maxCombo =
+      this.vm.result.combo > this.vm.result.maxCombo
+        ? this.vm.result.combo
+        : this.vm.result.maxCombo;
+    this.vm.fever.percent += (1 - percentage) / 100;
+    this.hitIndicator(this.vm);
+  }
+
+  isOutOfCanvas() {
+    const yOut = this.y > this.canvas.height;
+    const isOut =
+      (yOut && !this.isHoldNote) ||
+      (yOut &&
+        this.isHoldNote &&
+        !this.isUserHolding &&
+        !this.isHoldNoteFinished(true));
+    if (this.vm.started && isOut && !this.noteFailed) {
+      this.vm.result.marks.miss += 1;
+      this.vm.result.totalHitNotes += 1;
+      this.vm.result.combo = 0;
+      this.vm.markJudge = "Miss";
+      this.vm.fever.percent -= 0.1;
       this.vibrate([20, 20, 50]);
-      this.hitIndicator(vm);
+      this.hitIndicator(this.vm);
+      this.noteFailed = true;
     }
-    return isOut;
-  };
+    const shouldClean =
+      (yOut && !this.isHoldNote) ||
+      (this.isHoldNote &&
+        (this.holdNoteY > this.canvas.height || this.isHoldingDone));
+    return shouldClean;
+  }
+
+  isHoldNoteFinished(nearly) {
+    const offset = nearly
+      ? Math.min.apply(0, [this.holdNoteHeight / 2, 100])
```

Comment:
```suggestion
      ? Math.min.apply(0, [this.holdNoteHeight / 2, 100])
```

What does this call do again? I understand that `apply` is prototype-inherited from `Object` and invokes a function call, but I'm not sure what this is doing?

Reply:
See [Math.min.apply(0, array) - why?](https://stackoverflow.com/questions/2870015/math-min-apply0-array-why)

Reply:
Basically I'm getting either half of the long note's length or 100 px, whichever is shorter gets chosen, to check whether the player releases the hold note at an appropriate timing or not.